Add implementation of hashCode method for Dynamic Value along with the corresponding test

BijenderKumar1 opened the pull request:
/claim #655
In order to resolve the issue of incosistent `hashCode` values for DynamicValue objects, I have added an implementation of the `hashCode` method for the `DynamicValue` class. I have also added a test to ensure that the `hashCode` values are consistent. Thanks!

BijenderKumar1 commented:
@jdegoes This is ready for review. Thanks!

jdegoes commented:
Unfortunately, the job is bigger than just hash code. You also need to fix equality. In addition, the test suite needs to be more comprehensive, verifying that different ways of creating equivalent dynamic value compare and hash equal. 

I will increase the bounty on this one as there is more work than I thought.

BijenderKumar1 commented:
@jdegoes I have fixed equality as well by overriding the `equals` method for `DynamicValue`. I have also expanded the text suite to incorporate checks for different ways of creating equivalent dynamic values. Thanks!

BijenderKumar1 commented:
@jdegoes Just a gentle reminder to please review this! Thanks!

BijenderKumar1 commented:
@jdegoes Really sorry to bother you! Just a gentle reminder. Thanks!

BijenderKumar1 commented:
@jdegoes Just a gentle reminder! Thanks!

jdegoes commented on tests/shared/src/test/scala-2/zio/schema/DynamicValueSpec.scala:
```diff
@@ -8,6 +8,18 @@ import zio.test.{ Sized, TestConfig, _ }
 
 object DynamicValueSpec extends ZIOSpecDefault {
 
+  case class Person(name: String, age: Int)
```
You should not need any classes at all, because you are testing `DynamicValue`. You should create all DynamicValue manually.

BijenderKumar1 replied:
Thanks! Resolved in the next revision.

jdegoes commented on tests/shared/src/test/scala-2/zio/schema/DynamicValueSpec.scala:
```diff
@@ -100,6 +112,33 @@ object DynamicValueSpec extends ZIOSpecDefault {
+      ),
+      suite("hashCode and equality consistency")(
+        test("hashCode does not change") {
+          val primitive1 = DynamicValue.Primitive(123, StandardType.IntType)
```
Rather than these tests, which test concrete values, you should use a generator for dynamic value `Gen`. I am not sure if one exists, but if not, you can create one pretty easily.

BijenderKumar1 replied:
Thanks! Resolved in the next revision.

jdegoes commented:
@BijenderKumar1 The tests are not testing anything robust.

A robust test will be property-based, using dynamically generated `DynamicValue`, without involvement of `Schema` or specific classes, and will test for consistent hash and equality.

BijenderKumar1 commented:
@jdegoes  Thanks for the review. I have added a new function to dynamically generate  `DynamicValue`s and added tests for consistent hash and equality. Thanks!

BijenderKumar1 commented:
@jdegoes Just a gentle reminder to review this. Thanks!

BijenderKumar1 commented:
@jdegoes Just a gentle reminder to review this. Thanks!

BijenderKumar1 commented:
@jdegoes Just a gentle reminder to review this. Really sorry to bother you repeatedly! Thanks!

BijenderKumar1 commented:
@jdegoes Hope you are doing well! Just a gentle reminder for reviewing this. Thanks!

BijenderKumar1 commented:
@jdegoes @987Nabil Just a gentle reminder to please review this. I will be really thankful!

987Nabil commented:
@BijenderKumar1 There are still issues. I'll write once I have the time for it

BijenderKumar1 commented:
@987Nabil No worries! I will be super happy to incorporate your suggestions. Thanks!

BijenderKumar1 commented:
@jdegoes Hope you are doing well! Just a gentle reminder to please review this. Thanks!
